rust/basic_ethereum/src/ethereum_wallet.rs

```diff
@@ -0,0 +1,113 @@
+//! A demo of a very bare-bones ethereum "wallet".
+//!
+//! The wallet here showcases how ethereum addresses can be computed
+//! and how ethereum transactions can be signed. It is missing several
+//! pieces that any production-grade wallet would have, such as error handling, access-control, caching, etc.
+
+use crate::ecdsa::EcdsaPublicKey;
+use crate::state::{lazy_call_ecdsa_public_key, read_state};
+use candid::Principal;
+use ic_crypto_ecdsa_secp256k1::{PublicKey, RecoveryId};
+use ic_ethereum_types::Address;
+use serde_bytes::ByteBuf;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct EthereumWallet {
+    owner: Principal,
+    derived_public_key: EcdsaPublicKey,
+}
+
+impl AsRef<PublicKey> for EthereumWallet {
+    fn as_ref(&self) -> &PublicKey {
+        self.derived_public_key.as_ref()
+    }
+}
+
+impl EthereumWallet {
+    pub async fn new(owner: Principal) -> Self {
+        let derived_public_key = derive_public_key(&owner, &lazy_call_ecdsa_public_key().await);
+        Self {
+            owner,
+            derived_public_key,
+        }
+    }
+
+    pub fn ethereum_address(&self) -> Address {
+        Address::from(&self.derived_public_key)
+    }
+
+    pub async fn sign_with_ecdsa(&self, message_hash: [u8; 32]) -> ([u8; 64], RecoveryId) {
+        use ic_cdk::api::management_canister::ecdsa::SignWithEcdsaArgument;
+
+        let derivation_path = derivation_path(&self.owner);
+        let key_id = read_state(|s| s.ecdsa_key_id());
+        let (result,) =
+            ic_cdk::api::management_canister::ecdsa::sign_with_ecdsa(SignWithEcdsaArgument {
+                message_hash: message_hash.to_vec(),
+                derivation_path,
+                key_id,
+            })
+            .await
+            .expect("failed to sign with ecdsa");
+        let signature_length = result.signature.len();
```

Comment:
Since `signature_length` is only used in the `panic!` message, how about just putting `result.signature.len()` there?

Reply:
This you cannot because `<[u8; 64]>::try_from(result.signature)` consumes (takes ownership of) `result.signature`.  So doing it that way would require an additional cloning, which seems worse than this variable declaration.

Reply:
I see. Thanks for the explanation!